File: 101/scala/src/test/scala/HashtableTest.scala
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class HashtableTest extends AnyFunSuite with Matchers {
  val Key1 = "Key1"
  val Key2 = "Key2"
  val Value1 = "Value1"
  val Value2 = "Value2"

  test("it should add elements") {
    val hash = new Hashtable[String, String](10)
    hash.put(Key1, Value1)
    hash.put(Key2, Value2)
    hash.get(Key1) should be(Some(Value1))
    hash.get(Key2) should be(Some(Value2))
    hash.get("key").isEmpty should be(true)
  }

  test("it should overwrite elements") {
    val hash = new Hashtable[String, String](10)
    hash.put(Key1, Value1)
    hash.put(Key1, Value2)
    hash.get(Key1) should be(Some(Value2))
  }

  test("it should calculate size") {
    val hash = new Hashtable[String, String](10)
    hash.put(Key1, Value1)
    hash.put(Key2, Value2)
    hash.size should be(2)
  }

  test("it should get keys") {
    val hash = new Hashtable[String, String](10)
    hash.put(Key1, Value1)
    hash.put(Key2, Value2)
    hash.keys.toSeq.sorted should be(Seq(Key1, Key2))
  }

  test("it should get values") {
    val hash = new Hashtable[String, String](10)
    hash.put(Key1, Value1)
    hash.put(Key2, Value2)
    hash.values.toSeq.sorted should be(Seq(Value1, Value2))
  }

  test("it should remove key") {
    val hash = new Hashtable[String, String](10)
    hash.put(Key1, Value1)
    hash.get(Key1) should be(Some(Value1))
    hash.remove(Key1)
    hash.get(Key1) should be(None)
  }
}

File: 101/ts/test/hashTable.test.ts
import {HashTable} from '../src/hashTable';

describe('HashTable', () => {
  const key1 = 'key1';
  const key2 = 'key2';
  const value1 = 'value1';
  const value2 = 'value2';

  const static_hash = new HashTable(10);
  static_hash.put(key1, value1);
  static_hash.put(key2, value2);

  test('should add elements', () => {
    const hash = new HashTable(10);
    hash.put(key1, value1);
    hash.put(key2, value2);
    expect(hash.get(key1)).toBe(value1);
    expect(hash.get(key2)).toBe(value2);
    expect(hash.get('key')).toBeUndefined();
  });

  test('should overwrite elements', () => {
    const hash = new HashTable(10);
    hash.put(key1, value1);
    hash.put(key1, value2);
    expect(hash.get(key1)).toBe(value2);
  });

  test('should calculate size', () => {
    expect(static_hash.size()).toBe(2);
  });

  test('should get keys', () => {
    expect(static_hash.keys()).toStrictEqual([key1, key2]);
  });

  test('should get values', () => {
    expect(static_hash.values()).toEqual([value1, value2]);
  });

  test('should remove key', () => {
    const hash = new HashTable(10);
    hash.put(key1, value1);
    expect(hash.get(key1)).toBe(value1);
    hash.remove(key1);
    expect(hash.get(key1)).toBeUndefined();
  });
});
